libraries/AP_GPS/AP_GPS.cpp

```diff
@@ -772,72 +827,82 @@ void AP_GPS::update(void)
-                    // choose GPS with highest state or higher number of satellites
-                    if ((state[i].status > state[primary_instance].status) ||
-                        ((state[i].status == state[primary_instance].status) && (state[i].num_sats > state[primary_instance].num_sats))) {
-                        primary_instance = i;
-                        _last_instance_swap_ms = now;
-                    }
-                }
-            } else {
-                // handle switch between real GPSs
-                for (uint8_t i=0; i<GPS_MAX_RECEIVERS; i++) {
-                    if (i == primary_instance) {
-                        continue;
-                    }
-                    if (state[i].status > state[primary_instance].status) {
-                        // we have a higher status lock, or primary is set to the blended GPS, change GPS
-                        primary_instance = i;
-                        _last_instance_swap_ms = now;
-                        continue;
-                    }
-
-                    bool another_gps_has_1_or_more_sats = (state[i].num_sats >= state[primary_instance].num_sats + 1);
-
-                    if (state[i].status == state[primary_instance].status && another_gps_has_1_or_more_sats) {
-
-                        bool another_gps_has_2_or_more_sats = (state[i].num_sats >= state[primary_instance].num_sats + 2);
-
-                        if ((another_gps_has_1_or_more_sats && (now - _last_instance_swap_ms) >= 20000) ||
-                            (another_gps_has_2_or_more_sats && (now - _last_instance_swap_ms) >= 5000)) {
-                            // this GPS has more satellites than the
-                            // current primary, switch primary. Once we switch we will
-                            // then tend to stick to the new GPS as primary. We don't
-                            // want to switch too often as it will look like a
-                            // position shift to the controllers.
-                            primary_instance = i;
-                            _last_instance_swap_ms = now;
-                        }
-                    }
-                }
+        return;
+    }
+
+    if (_auto_switch == 0) {
+        // AUTO_SWITCH is 0 so no switching of GPSs, always use first instance
+        primary_instance = 0;
+        return;
+    }
+
+    if (_auto_switch == 3) {
+        // always select the second GPS instance
+        primary_instance = 1;
+        return;
+    }
+
+    uint32_t now = AP_HAL::millis();
+
+    // special handling of RTK moving baseline pair. If a rover has a
+    // RTK fixed lock and yaw available then always select it as
+    // primary. This ensures that the yaw data and position/velocity
+    // data is time aligned whenever we provide yaw to the EKF
+    for (uint8_t i=0; i<GPS_MAX_RECEIVERS; i++) {
+        if (_type[i] == GPS_TYPE_UBLOX_RTK_ROVER &&
+            state[i].status == GPS_OK_FIX_3D_RTK_FIXED &&
```

Comment:
it is deliberate. I was concerned that have_gps_yaw may not be unset yet. We don't unset it will we get a new RELPOSND, but this code selects GPS before than